sklearn/preprocessing/tests/test_data.py

```diff
@@ -851,6 +854,298 @@ def test_robust_scaler_iris_quantiles():
+    X_neg = sparse.csc_matrix(X_neg)
+
+    assert_raises_regex(ValueError, "Invalid value for 'n_quantiles'",
+                        QuantileTransformer(n_quantiles=0).fit, X_neg)
+    assert_raises_regex(ValueError, "Invalid value for 'subsample'",
+                        QuantileTransformer(subsample=0).fit, X_neg)
+    assert_raises_regex(ValueError, "Invalid value for 'smoothing_noise'",
+                        QuantileTransformer(smoothing_noise=0).fit, X_neg)
+
+    transformer = QuantileTransformer(n_quantiles=10)
+    assert_raises_regex(ValueError, "QuantileTransformer only accepts "
+                        "non-negative sparse matrices", transformer.fit, X_neg)
+    transformer.fit(X)
+    assert_raises_regex(ValueError, "QuantileTransformer only accepts "
+                        "non-negative sparse matrices",
+                        transformer.transform, X_neg)
+    assert_raises_regex(ValueError, "QuantileTransformer only accepts "
+                        "non-negative sparse matrices",
+                        transformer.inverse_transform, X_neg)
+
+    X_bad_feat = np.array([[0, 25, 50, 0, 0, 0, 75, 0, 0, 100],
+                           [0, 0, 2.6, 4.1, 0, 0, 2.3, 0, 9.5, 0.1]]).T
+    assert_raises_regex(ValueError, "X does not have the same number of "
+                        "feature than the previously fitted data.",
+                        transformer.transform, X_bad_feat)
+    assert_raises_regex(ValueError, "X does not have the same number of "
+                        "feature than the previously fitted data.",
+                        transformer.inverse_transform, X_bad_feat)
+
+    assert_raises_regex(ValueError, "'output_distribution' has to be either"
+                        " 'norm' or 'uniform'. Got rnd instead.",
+                        QuantileTransformer(
+                            n_quantiles=10,
+                            output_distribution='rnd').fit_transform, X)
+    assert_raises_regex(ValueError, "'output_distribution' has to be either"
+                        " 'norm' or 'uniform'. Got rnd instead.",
+                        QuantileTransformer(n_quantiles=10,
+                                            output_distribution='rnd').fit(
+                            X).inverse_transform, X)
+
+
+def test_quantile_transform_ignore_zeros():
+    X = np.array([[0, 0, 0, 0, 0],
+                  [1, 0, 2, 2, 1]]).T
+    X_sparse = sparse.csc_matrix(X)
+    transformer = QuantileTransformer(ignore_implicit_zeros=True,
+                                      n_quantiles=5)
+
+    # dense case -> warning raise
+    assert_warns_message(UserWarning, "'ignore_implicit_zeros' takes effect"
+                         " only with sparse matrix. This parameter has no"
+                         " effect.", transformer.fit, X)
+
+    X_gt = np.array([[0, 0, 0, 0, 0],
+                     [0, 0, 1, 1, 0]]).T
+    X_trans = transformer.fit_transform(X_sparse)
+    assert_almost_equal(X_gt, X_trans.A)
```

Comment:
`csc_matrix`, among *some* other scipy sparse matrices, will faithfully store *explicit* zeros (i.e. there's an element with value 0 in csc_matrix.data). This can be used in applications where the sparse matrix represents a sparse mask (i.e. implicit element should be understood as inf or NaN or something) rather than a matrix that is sparsely nonzero. It's somewhat problematic that scipy does not have a separate representation for that case. But if some use of sparse matrices in scikit-learn is treating it as a sparse mask (which ignoring implicit zeros is), and zero is a valid data element in that case, we should test that it is handled correctly.

Is that helpful, or am I talking on the wrong point?